File: go/internal/runtime/observer.go
// Package runtime provides system observability (vitals, topology, health).
// Implements DIP: Observer depends on Runtime interface, not concrete Docker.
package runtime

import (
	"context"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/joss/urp/internal/graph"
)

// ContainerState represents container metrics (Φ energy).
type ContainerState struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Status      string  `json:"status"`
	CPUPercent  float64 `json:"cpu_percent"`
	MemoryBytes int64   `json:"memory_bytes"`
	MemoryLimit int64   `json:"memory_limit"`
	MemoryPct   float64 `json:"memory_pct"`
	NetworkRx   int64   `json:"network_rx"`
	NetworkTx   int64   `json:"network_tx"`
}

// HealthIssue represents a container health problem (⊥ conflict).
type HealthIssue struct {
	Container string `json:"container"`
	Type      string `json:"type"`
	Severity  string `json:"severity"`
	Detail    string `json:"detail"`
}

// NetworkTopology represents container network layout (⊆ inclusion).
type NetworkTopology struct {
	Containers []ContainerInfo `json:"containers"`
	Networks   []string        `json:"networks"`
	Error      string          `json:"error,omitempty"`
}

// ContainerInfo has basic container details.
type ContainerInfo struct {
	ID       string   `json:"id"`
	Name     string   `json:"name"`
	Image    string   `json:"image"`
	Ports    []string `json:"ports"`
	Networks []string `json:"networks"`
}

// Observer provides runtime observation capabilities.
// Uses Runtime interface for DIP - can be injected with mock for testing.
type Observer struct {
	db      graph.Driver
	runtime Runtime // DIP: depends on interface, not concrete Docker
}

// ObserverOption configures Observer via functional options (DIP).
type ObserverOption func(*Observer)

// WithRuntime injects a custom runtime (for testing).
func WithRuntime(r Runtime) ObserverOption {
	return func(o *Observer) { o.runtime = r }
}

// NewObserver creates an observer, auto-detecting container runtime.
func NewObserver(db graph.Driver, opts ...ObserverOption) *Observer {
	o := &Observer{
		db:      db,
		runtime: DetectRuntime(), // Default: auto-detect
	}

	for _, opt := range opts {
		opt(o)
	}

	return o
}

// Vitals returns current container states.
func (o *Observer) Vitals(ctx context.Context) ([]ContainerState, error) {
	if o.runtime == nil || !o.runtime.Available() {
		return nil, fmt.Errorf("Docker not found. Install Docker to use URP containers.")
	}

	// Use Runtime interface (DIP)
	rawStats, err := o.runtime.Stats(ctx)
	if err != nil {
		return nil, fmt.Errorf("stats failed: %w", err)
	}

	var states []ContainerState
	for _, raw := range rawStats {
		state := ContainerState{
			ID:   truncateID(raw.ID),
			Name: raw.Name,
		}

		// Parse CPU (e.g., "1.23%")
		cpuStr := strings.TrimSuffix(raw.CPUPercent, "%")
		state.CPUPercent, _ = strconv.ParseFloat(cpuStr, 64)

		// Parse memory (e.g., "123MiB / 1GiB")
		state.MemoryBytes, state.MemoryLimit = parseMemUsage(raw.MemUsage)
		if state.MemoryLimit > 0 {
			state.MemoryPct = float64(state.MemoryBytes) / float64(state.MemoryLimit) * 100
		}

		// Parse network (e.g., "1.2kB / 3.4kB")
		state.NetworkRx, state.NetworkTx = parseNetIO(raw.NetIO)

		state.Status = "running"
		states = append(states, state)

		// Store in graph if connected
		if o.db != nil {
			o.storeVitals(ctx, state)
		}
	}

	return states, nil
}

// truncateID truncates container ID to 12 characters.
func truncateID(id string) string {
	if len(id) > 12 {
		return id[:12]
	}
	return id
}

func parseMemUsage(s string) (used, limit int64) {
	// Format: "123MiB / 1GiB" or "123.4MB / 1GB"
	parts := strings.Split(s, " / ")
	if len(parts) == 2 {
		used = parseSize(parts[0])
		limit = parseSize(parts[1])
	}
	return
}

func parseSize(s string) int64 {
	s = strings.TrimSpace(s)

	// Match number and unit
	re := regexp.MustCompile(`([\d.]+)([A-Za-z]+)`)
	matches := re.FindStringSubmatch(s)
	if len(matches) != 3 {
		return 0
	}

	val, _ := strconv.ParseFloat(matches[1], 64)
	unit := strings.ToUpper(matches[2])

	multipliers := map[string]float64{
		"B":   1,
		"KB":  1024,
		"KIB": 1024,
		"MB":  1024 * 1024,
		"MIB": 1024 * 1024,
		"GB":  1024 * 1024 * 1024,
		"GIB": 1024 * 1024 * 1024,
	}

	if mult, ok := multipliers[unit]; ok {
		return int64(val * mult)
	}
	return int64(val)
}

func parseNetIO(s string) (rx, tx int64) {
	// Format: "1.2kB / 3.4kB"
	parts := strings.Split(s, " / ")
	if len(parts) == 2 {
		rx = parseSize(parts[0])
		tx = parseSize(parts[1])
	}
	return
}

func (o *Observer) storeVitals(ctx context.Context, state ContainerState) {
	query := `
		MERGE (c:Container {id: $id})
		SET c.name = $name,
		    c.status = $status,
		    c.cpu_phi = $cpu,
		    c.mem_bytes = $mem,
		    c.mem_limit = $mem_limit,
		    c.mem_percent = $mem_pct,
		    c.net_rx = $rx,
		    c.net_tx = $tx,
		    c.last_seen = $ts
	`

	o.db.ExecuteWrite(ctx, query, map[string]any{
		"id":        state.ID,
		"name":      state.Name,
		"status":    state.Status,
		"cpu":       state.CPUPercent,
		"mem":       state.MemoryBytes,
		"mem_limit": state.MemoryLimit,
		"mem_pct":   state.MemoryPct,
		"rx":        state.NetworkRx,
		"tx":        state.NetworkTx,
		"ts":        time.Now().Unix(),
	})
}

// Topology returns container network topology.
func (o *Observer) Topology(ctx context.Context) (*NetworkTopology, error) {
	result := &NetworkTopology{
		Containers: []ContainerInfo{},
		Networks:   []string{},
	}

	if o.runtime == nil || !o.runtime.Available() {
		result.Error = "no container runtime found"
		return result, nil
	}

	// Use Runtime interface (DIP)
	rawContainers, err := o.runtime.List(ctx)
	if err != nil {
		result.Error = fmt.Sprintf("list failed: %v", err)
		return result, nil
	}

	networkSet := make(map[string]bool)

	for _, raw := range rawContainers {
		info := ContainerInfo{
			ID:    truncateID(raw.ID),
			Name:  raw.Name,
			Image: raw.Image,
		}

		// Parse ports
		if raw.Ports != "" {
			info.Ports = strings.Split(raw.Ports, ", ")
		}

		// Parse networks
		if raw.Networks != "" {
			info.Networks = strings.Split(raw.Networks, ",")
			for _, n := range info.Networks {
				networkSet[strings.TrimSpace(n)] = true
			}
		}

		result.Containers = append(result.Containers, info)

		// Store in graph
		if o.db != nil {
			o.storeTopology(ctx, info)
		}
	}

	for net := range networkSet {
		result.Networks = append(result.Networks, net)
	}

	return result, nil
}

func (o *Observer) storeTopology(ctx context.Context, info ContainerInfo) {
	for _, net := range info.Networks {
		query := `
			MERGE (c:Container {id: $cid})
			SET c.name = $name, c.image = $image
			MERGE (n:Network {name: $net})
			MERGE (c)-[:CONNECTED_TO]->(n)
		`

		o.db.ExecuteWrite(ctx, query, map[string]any{
			"cid":   info.ID,
			"name":  info.Name,
			"image": info.Image,
			"net":   strings.TrimSpace(net),
		})
	}
}

// Health checks for container issues.
func (o *Observer) Health(ctx context.Context) ([]HealthIssue, error) {
	var issues []HealthIssue

	if o.runtime == nil || !o.runtime.Available() {
		issues = append(issues, HealthIssue{
			Container: "system",
			Type:      "NO_RUNTIME",
			Severity:  "ERROR",
			Detail:    "Docker not found. Install Docker to use URP.",
		})
		return issues, nil
	}

	// Use Runtime interface (DIP)
	rawContainers, err := o.runtime.ListAll(ctx)
	if err != nil {
		issues = append(issues, HealthIssue{
			Container: o.runtime.Name(),
			Type:      "CMD_FAILED",
			Severity:  "ERROR",
			Detail:    err.Error(),
		})
		return issues, nil
	}

	for _, raw := range rawContainers {
		name := raw.Name
		status := strings.ToLower(raw.Status)
		state := strings.ToLower(raw.State)

		// Check for restart loops
		if strings.Contains(status, "restarting") {
			issues = append(issues, HealthIssue{
				Container: name,
				Type:      "RESTART_LOOP",
				Severity:  "ERROR",
				Detail:    status,
			})
		}

		// Check for exited containers with error
		if state == "exited" {
			if strings.Contains(status, "exited (0)") {
				continue // Normal exit
			}
			issues = append(issues, HealthIssue{
				Container: name,
				Type:      "EXIT_ERROR",
				Severity:  "ERROR",
				Detail:    status,
			})
		}

		// Check for unhealthy
		if strings.Contains(status, "unhealthy") {
			issues = append(issues, HealthIssue{
				Container: name,
				Type:      "UNHEALTHY",
				Severity:  "ERROR",
				Detail:    status,
			})
		}
	}

	return issues, nil
}

// Runtime returns detected container runtime name.
func (o *Observer) RuntimeName() string {
	if o.runtime == nil {
		return ""
	}
	return o.runtime.Name()
}

// RuntimeAvailable returns true if a container runtime is available.
func (o *Observer) RuntimeAvailable() bool {
	return o.runtime != nil && o.runtime.Available()
}

File: legacy/python/observer.py
"""
Container/Runtime Observer - Propiocepción del sistema.
Φ (Morphism): CPU/RAM as energy flow
D (Domain): Container existence and boundaries
τ (Vector): Log events as temporal sequence
⊥ (Orthogonal): Health status, conflicts
"""
import os
import re
import time
import json
from dataclasses import dataclass
from typing import Iterator, Optional
from datetime import datetime

try:
    import docker
    HAS_DOCKER = True
except ImportError:
    HAS_DOCKER = False


@dataclass
class ContainerState:
    """Current state of a container (Φ energy levels)."""
    id: str
    name: str
    status: str
    cpu_percent: float
    memory_bytes: int
    memory_limit: int
    network_rx: int
    network_tx: int


@dataclass
class LogEvent:
    """A parsed log event (τ temporal node)."""
    container_id: str
    timestamp: datetime
    level: str  # INFO, WARN, ERROR, FATAL
    message: str
    source: str


class ContainerObserver:
    """Observes Docker containers and feeds state into the graph."""

    # Patterns to detect log severity
    LOG_PATTERNS = [
        (r'\b(FATAL|PANIC|CRITICAL)\b', 'FATAL'),
        (r'\b(ERROR|ERR|EXCEPTION|FAIL)\b', 'ERROR'),
        (r'\b(WARN|WARNING)\b', 'WARN'),
        (r'\b(INFO|DEBUG|TRACE)\b', 'INFO'),
    ]

    def __init__(self, db):
        self._db = db
        self._client = None
        self._docker_error = None

        if not HAS_DOCKER:
            self._docker_error = "Docker SDK not installed. Run: pip install docker"
            return

        try:
            self._client = docker.from_env()
            # Test connection
            self._client.ping()
        except Exception as e:
            self._docker_error = f"Docker not accessible: {e}"
            self._client = None

    def snapshot_all(self) -> list[ContainerState]:
        """
        Take a snapshot of all running containers.
        Returns list of container states.
        """
        if not self._client:
            return []

        states = []
        for container in self._client.containers.list():
            state = self._get_container_state(container)
            if state:
                states.append(state)
                self._store_state(state)
        return states

    def get_error(self) -> Optional[str]:
        """Return Docker connection error if any."""
        return self._docker_error

    def _get_container_state(self, container) -> Optional[ContainerState]:
        """Extract energy metrics from a container."""
        try:
            stats = container.stats(stream=False)
            cpu = self._calculate_cpu_percent(stats)
            mem = stats['memory_stats'].get('usage', 0)
            mem_limit = stats['memory_stats'].get('limit', 1)

            networks = stats.get('networks', {})
            rx = sum(n.get('rx_bytes', 0) for n in networks.values())
            tx = sum(n.get('tx_bytes', 0) for n in networks.values())

            return ContainerState(
                id=container.short_id,
                name=container.name,
                status=container.status,
                cpu_percent=cpu,
                memory_bytes=mem,
                memory_limit=mem_limit,
                network_rx=rx,
                network_tx=tx,
            )
        except Exception as e:
            print(f"Warning: Could not get stats for {container.name}: {e}")
            return None

    def _calculate_cpu_percent(self, stats: dict) -> float:
        """Calculate CPU percentage from Docker stats."""
        try:
            cpu_delta = stats['cpu_stats']['cpu_usage']['total_usage'] - \
                        stats['precpu_stats']['cpu_usage']['total_usage']
            system_delta = stats['cpu_stats']['system_cpu_usage'] - \
                           stats['precpu_stats']['system_cpu_usage']
            num_cpus = len(stats['cpu_stats']['cpu_usage'].get('percpu_usage', [1]))
            if system_delta > 0:
                return (cpu_delta / system_delta) * num_cpus * 100.0
        except (KeyError, ZeroDivisionError):
            pass
        return 0.0

    def _store_state(self, state: ContainerState):
        """Store container state in graph (D + Φ)."""
        self._db.execute_query("""
            MERGE (c:Container {id: $id})
            SET c.name = $name,
                c.status = $status,
                c.cpu_phi = $cpu,
                c.mem_bytes = $mem,
                c.mem_limit = $mem_limit,
                c.mem_percent = $mem_pct,
                c.net_rx = $rx,
                c.net_tx = $tx,
                c.last_seen = $ts
        """, {
            "id": state.id,
            "name": state.name,
            "status": state.status,
            "cpu": round(state.cpu_percent, 2),
            "mem": state.memory_bytes,
            "mem_limit": state.memory_limit,
            "mem_pct": round(state.memory_bytes / max(state.memory_limit, 1) * 100, 2),
            "rx": state.network_rx,
            "tx": state.network_tx,
            "ts": int(time.time()),
        })

    def watch_logs(self, container_name: str, tail: int = 100) -> list[LogEvent]:
        """
        Capture recent logs and extract events (τ temporal).
        """
        if not self._client:
            return []

        try:
            container = self._client.containers.get(container_name)
            logs = container.logs(tail=tail, timestamps=True).decode('utf-8', errors='replace')
            events = list(self._parse_logs(container.short_id, logs))

            # Store error events in graph
            for event in events:
                if event.level in ('ERROR', 'FATAL'):
                    self._store_log_event(event)

            return events
        except docker.errors.NotFound:
            return []

    def _parse_logs(self, container_id: str, logs: str) -> Iterator[LogEvent]:
        """Parse log lines into structured events."""
        for line in logs.strip().split('\n'):
            if not line:
                continue

            # Try to extract timestamp (Docker format: 2024-01-01T00:00:00.000000000Z)
            ts_match = re.match(r'^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})', line)
            if ts_match:
                try:
                    ts = datetime.fromisoformat(ts_match.group(1))
                except ValueError:
                    ts = datetime.now()
                message = line[ts_match.end():].strip()
            else:
                ts = datetime.now()
                message = line

            # Detect severity
            level = 'INFO'
            for pattern, lvl in self.LOG_PATTERNS:
                if re.search(pattern, message, re.IGNORECASE):
                    level = lvl
                    break

            yield LogEvent(
                container_id=container_id,
                timestamp=ts,
                level=level,
                message=message[:500],  # Truncate
                source='docker',
            )

    def _store_log_event(self, event: LogEvent):
        """Store significant log events in graph."""
        self._db.execute_query("""
            MATCH (c:Container {id: $cid})
            CREATE (e:LogEvent {
                level: $level,
                message: $msg,
                timestamp: $ts
            })
            CREATE (c)-[:EMITTED]->(e)
        """, {
            "cid": event.container_id,
            "level": event.level,
            "msg": event.message,
            "ts": event.timestamp.isoformat(),
        })

    def get_topology(self) -> dict:
        """
        Get container network topology (⊆ inclusion).
        Returns dict of containers and their network connections.
        """
        topology = {"containers": [], "networks": [], "error": self._docker_error}

        if not self._client:
            return topology

        for container in self._client.containers.list():
            info = container.attrs
            nets = info.get('NetworkSettings', {}).get('Networks', {})

            c_info = {
                "id": container.short_id,
                "name": container.name,
                "image": info.get('Config', {}).get('Image', ''),
                "ports": list(info.get('NetworkSettings', {}).get('Ports', {}).keys()),
                "networks": list(nets.keys()),
                "mounts": [m['Destination'] for m in info.get('Mounts', [])],
            }
            topology["containers"].append(c_info)

            for net_name, net_info in nets.items():
                if net_name not in topology["networks"]:
                    topology["networks"].append(net_name)

                # Store topology in graph
                self._db.execute_query("""
                    MERGE (c:Container {id: $cid})
                    MERGE (n:Network {name: $net})
                    MERGE (c)-[:CONNECTED_TO {ip: $ip}]->(n)
                """, {
                    "cid": container.short_id,
                    "net": net_name,
                    "ip": net_info.get('IPAddress', ''),
                })

        return topology

    def check_health(self) -> list[dict]:
        """
        Check container health status (⊥ orthogonal conflicts).
        Returns list of health issues.
        """
        issues = []

        if not self._client:
            if self._docker_error:
                issues.append({"container": "docker", "issue": "unavailable", "detail": self._docker_error})
            return issues

        for container in self._client.containers.list(all=True):
            state = container.attrs.get('State', {})
            health = state.get('Health', {})

            # Restart loops
            restart_count = state.get('RestartCount', 0)
            if restart_count > 3:
                issues.append({
                    "container": container.name,
                    "type": "RESTART_LOOP",
                    "severity": "ERROR",
                    "detail": f"Restarted {restart_count} times",
                })

            # Health check failures
            if health.get('Status') == 'unhealthy':
                issues.append({
                    "container": container.name,
                    "type": "UNHEALTHY",
                    "severity": "ERROR",
                    "detail": health.get('Log', [{}])[-1].get('Output', '')[:200],
                })

            # OOMKilled
            if state.get('OOMKilled'):
                issues.append({
                    "container": container.name,
                    "type": "OOM_KILLED",
                    "severity": "FATAL",
                    "detail": "Container killed by OOM",
                })

            # Exited with error
            if state.get('Status') == 'exited' and state.get('ExitCode', 0) != 0:
                issues.append({
                    "container": container.name,
                    "type": "EXIT_ERROR",
                    "severity": "ERROR",
                    "detail": f"Exit code: {state.get('ExitCode')}",
                })

        return issues


class ProcessObserver:
    """Observes host processes when not using Docker."""

    def __init__(self, db):
        self._db = db

    def snapshot_processes(self, filter_cmd: Optional[str] = None) -> list[dict]:
        """Get process info from /proc (Linux) or ps."""
        import subprocess

        try:
            result = subprocess.run(
                ['ps', 'aux', '--no-headers'],
                capture_output=True,
                text=True,
                timeout=5,
            )

            processes = []
            for line in result.stdout.strip().split('\n'):
                parts = line.split(None, 10)
                if len(parts) >= 11:
                    cmd = parts[10]
                    if filter_cmd and filter_cmd not in cmd:
                        continue
                    processes.append({
                        "user": parts[0],
                        "pid": parts[1],
                        "cpu": float(parts[2]),
                        "mem": float(parts[3]),
                        "command": cmd[:200],
                    })

            return processes
        except Exception as e:
            return [{"error": str(e)}]
